src/wired/dataclasses/decorators.py

```diff
@@ -0,0 +1,59 @@
+from venusian import attach, Scanner
+from wired import ServiceRegistry
+
+from .registration import register_dataclass
+
+
+# noinspection PyPep8Naming
+class singleton:
+    def __init__(self,
+                 for_=None,
+                 context=None,
+                 name: str = None,
+                 ):
+        self.for_ = for_
+        self.context = context if context else for_
+        self.name = name
+
+    def __call__(self, wrapped):
+        def callback(scanner: Scanner, name: str, cls):
+            registry: ServiceRegistry = getattr(scanner, 'registry')
+
+            # Later we can do some dataclass field sniffing on singletons
+            instance = cls()
+
+            # If there is a for_ use it, otherwise, register for the same
+            # class as the instance
+            for_ = self.for_ if self.for_ else cls
+            registry.register_singleton(instance, for_)
+
+        attach(wrapped, callback)
+        return wrapped
+
+
+# noinspection PyPep8Naming
+class factory:
+    def __init__(self,
+                 for_=None,
+                 context=None,
+                 name: str = None,
+                 ):
+        self.for_ = for_
+        self.context = context if context else for_
+        self.name = name
+
+    def __call__(self, wrapped):
+        def callback(scanner: Scanner, name: str, cls):
+            registry: ServiceRegistry = getattr(scanner, 'registry')
+            # If there is a for_ use it, otherwise, register for the same
+            # class as the instance
+            for_ = self.for_ if self.for_ else cls
+            register_dataclass(
+                registry,
+                cls,
+                for_=for_,
+                context=self.context
+            )
+
+        attach(wrapped, callback)
```

Comment:
`category='wired'`

Reply:
Hmm, interesting, I had a feeling `category` was something I should know about. Can you explain the impact?

Reply:
It basically allows various libraries to only scan decorators they care about. It matters who is scanning the module (pyramid's configurator, versus this custom scanner, versus some other scanner) as each decorator expects certain attributes on the scanner. Thus you want to give the user control over which decorators apply to this scan. See also the default for the scanner is to pick up all categories which can exacerbate the problem.

https://github.com/Pylons/venusian/blob/72f46814884372cf70b343ef8bb5bdcb9c5919c5/venusian/__init__.py#L29-L32